File: backend/payment-service/src/course-client/auth-service.client.ts
import {Injectable, NotFoundException, ServiceUnavailableException} from '@nestjs/common';
import {ConfigService} from '@nestjs/config';

export interface UserContact {
  id: string;
  email: string;
  fullName: string | null;
}

interface UserApiResponse {
  success: boolean;
  data?: UserContact;
  error?: string;
}

// Hits auth-service's /api/internal/users/{id} — unauthenticated but never routed through the
// gateway (see InternalUserController in auth-service), so it's reachable only from other
// backend services. No X-User-Id/X-User-Role headers needed here, unlike CourseServiceClient.
@Injectable()
export class AuthServiceClient {
  constructor(private readonly config: ConfigService) {}

  async getUser(userId: string): Promise<UserContact> {
    const baseUrl = this.config.get<string>('AUTH_SERVICE_URL', 'http://localhost:9001');
    const res = await fetch(`${baseUrl}/api/internal/users/${userId}`);

    if (res.status === 404) {
      throw new NotFoundException(`User not found: ${userId}`);
    }
    if (!res.ok) {
      throw new ServiceUnavailableException('auth-service request failed');
    }

    const body = (await res.json()) as UserApiResponse;
    if (!body.success || !body.data) {
      throw new NotFoundException(`User not found: ${userId}`);
    }
    return body.data;
  }
}

File: backend/enrollment-service/src/main/java/com/grammarcetamol/enrollment/client/AuthServiceClient.java
package com.grammarcetamol.enrollment.client;

import com.grammarcetamol.enrollment.config.AppProperties;
import com.grammarcetamol.shared.dto.ApiResponse;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.http.client.SimpleClientHttpRequestFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestClient;

import java.util.UUID;

/**
 * Internal-only lookup for "who do we email" — hits auth-service's /api/internal/users/{id},
 * which is unauthenticated but never routed through the gateway (see
 * InternalUserController in auth-service), so it's reachable only from other backend services.
 * Unlike CourseServiceClient, no X-User-Id/X-User-Role headers are needed here since the target
 * endpoint doesn't check them.
 */
@Component
public class AuthServiceClient {

    public record UserContactDto(UUID id, String email, String fullName) {
    }

    private final RestClient restClient;

    public AuthServiceClient(AppProperties appProperties) {
        // Same SimpleClientHttpRequestFactory choice as CourseServiceClient — avoids the
        // Windows loopback-socket issue that java.net.http.HttpClient's NIO selector trips.
        this.restClient = RestClient.builder()
            .baseUrl(appProperties.getAuthServiceUrl())
            .requestFactory(new SimpleClientHttpRequestFactory())
            .build();
    }

    public UserContactDto getUser(UUID userId) {
        try {
            ApiResponse<UserContactDto> response = restClient.get()
                .uri("/api/internal/users/{id}", userId)
                .retrieve()
                .body(new org.springframework.core.ParameterizedTypeReference<>() {});
            if (response == null || response.getData() == null) {
                throw new EntityNotFoundException("User not found: " + userId);
            }
            return response.getData();
        } catch (HttpClientErrorException.NotFound e) {
            throw new EntityNotFoundException("User not found: " + userId);
        }
    }
}
